google/cloud/aiplatform/prediction/model_server.py

```diff
@@ -0,0 +1,190 @@
+    from fastapi import Request
+    from fastapi import Response
+except ImportError:
+    raise ImportError(
+        "FastAPI is not installed and is required to run model servers. "
+        'Please install the SDK using "pip install google-cloud-aiplatform[prediction]"'
+    )
+
+try:
+    import uvicorn
+except ImportError:
+    raise ImportError(
+        "Uvicorn is not installed and is required to run fastapi applications. "
+        'Please install the SDK using "pip install google-cloud-aiplatform[prediction]"'
+    )
+
+
+class CprModelServer:
+    """Model server to do custom prediction routines."""
+
+    def __init__(self):
+        """Initializes a fastapi application and sets the configs.
+
+        Args:
+            handler (Handler):
+                Required. The handler to handle requests.
+        """
+        self._init_logging()
+
+        if "HANDLER_MODULE" not in os.environ or "HANDLER_CLASS" not in os.environ:
+            raise ValueError(
+                "Both of the environment variables, HANDLER_MODULE and HANDLER_CLASS "
+                "need to be specified."
+            )
+        handler_module = importlib.import_module(os.environ.get("HANDLER_MODULE"))
+        handler_class = getattr(handler_module, os.environ.get("HANDLER_CLASS"))
+
+        predictor_class = None
+        if "PREDICTOR_MODULE" in os.environ:
+            predictor_module = importlib.import_module(
+                os.environ.get("PREDICTOR_MODULE")
+            )
+            predictor_class = getattr(
+                predictor_module, os.environ.get("PREDICTOR_CLASS")
+            )
+
+        self.handler = handler_class(
+            os.environ.get("AIP_STORAGE_URI"), predictor=predictor_class
+        )
+
+        if "AIP_HTTP_PORT" not in os.environ:
+            raise ValueError(
+                "The environment variable AIP_HTTP_PORT needs to be specified."
+            )
+        if (
+            "AIP_HEALTH_ROUTE" not in os.environ
+            or "AIP_PREDICT_ROUTE" not in os.environ
+        ):
+            raise ValueError(
+                "Both of the environment variables AIP_HEALTH_ROUTE and "
+                "AIP_PREDICT_ROUTE need to be specified."
+            )
+        self.http_port = int(os.environ.get("AIP_HTTP_PORT"))
+        self.health_route = os.environ.get("AIP_HEALTH_ROUTE")
+        self.predict_route = os.environ.get("AIP_PREDICT_ROUTE")
+
+        self.app = FastAPI()
+        self.app.add_api_route(
+            path=self.health_route,
+            endpoint=self.health,
+            methods=["GET"],
+        )
+        self.app.add_api_route(
+            path=self.predict_route,
+            endpoint=self.predict,
+            methods=["POST"],
+        )
+
+    async def __call__(self, scope, receive, send):
+        await self.app(scope, receive, send)
+
+    def _init_logging(self):
+        """Initializes the logging config."""
+        logging.basicConfig(
```

Comment:
Does model server run in the context of the user's workflow, or as its own standalone python process?

If its the former, this will override the user's own logging config and should be removed.

Reply:
The model server will run in our built new images for CPR instead of the users' environments.